src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,690 @@
+		}
+
+		void ProcessTypes (AssemblyDefinition assembly, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ()) {
+				XPathNavigator nav = iterator.Current;
+
+				string fullname = GetFullName (nav);
+
+				if (IsTypePattern (fullname)) {
+					ProcessTypePattern (fullname, assembly, nav);
+					continue;
+				}
+
+				TypeDefinition type = assembly.MainModule.GetType (fullname);
+
+				if (type == null) {
+					if (assembly.MainModule.HasExportedTypes) {
+						foreach (var exported in assembly.MainModule.ExportedTypes) {
+							if (fullname == exported.FullName) {
+								var resolvedExternal = exported.Resolve ();
+								if (resolvedExternal != null) {
+									type = resolvedExternal;
+									break;
+								}
+							}
+						}
+					}
+				}
+
+				if (type == null)
+					continue;
+
+				ProcessType (type, nav);
+			}
+		}
+
+		static bool IsTypePattern (string fullname)
+		{
+			return fullname.IndexOf ("*") != -1;
+		}
+
+		static Regex CreateRegexFromPattern (string pattern)
+		{
+			return new Regex (pattern.Replace (".", @"\.").Replace ("*", "(.*)"));
+		}
+
+		void MatchType (TypeDefinition type, Regex regex, XPathNavigator nav)
+		{
+			if (regex.Match (type.FullName).Success)
+				ProcessType (type, nav);
+
+			if (!type.HasNestedTypes)
+				return;
+
+			foreach (var nt in type.NestedTypes)
+				MatchType (nt, regex, nav);
+		}
+
+		void MatchExportedType (ExportedType exportedType, ModuleDefinition module, Regex regex, XPathNavigator nav)
+		{
+			if (regex.Match (exportedType.FullName).Success) {
+				TypeDefinition type = exportedType.Resolve ();
+				if (type != null) {
+					ProcessType (type, nav);
+				}
+			}
+		}
+
+
+		void ProcessTypePattern (string fullname, AssemblyDefinition assembly, XPathNavigator nav)
+		{
+			Regex regex = CreateRegexFromPattern (fullname);
+
+			foreach (TypeDefinition type in assembly.MainModule.Types) {
+				MatchType (type, regex, nav);
+			}
+
+			if (assembly.MainModule.HasExportedTypes) {
+				foreach (var exported in assembly.MainModule.ExportedTypes) {
+					MatchExportedType (exported, assembly.MainModule, regex, nav);
+				}
+			}
+		}
+
+		protected virtual void ProcessType (TypeDefinition type, XPathNavigator nav)
+		{
+#if !FEATURE_ILLINK
+			if (IsExcluded (nav))
+				return;
+#endif
+			ProcessTypeChildren (type, nav);
+
+			if (!type.HasNestedTypes)
+				return;
+
+			foreach (TypeDefinition nested in type.NestedTypes) {
+				var iterator = nav.SelectChildren ("type", _ns);
+				while (iterator.MoveNext ()) {
+					if (nested.Name == GetAttribute (iterator.Current, "name"))
+						ProcessTypeChildren (nested, iterator.Current);
+				}
+			}
+		}
+
+		void ProcessTypeChildren (TypeDefinition type, XPathNavigator nav)
+		{
+			if (nav.HasChildren) {
+				ProcessSelectedFields (nav, type);
+				ProcessSelectedMethods (nav, type);
+				ProcessSelectedProperties (nav, type);
+			}
+		}
+
+		void ProcessSelectedFields (XPathNavigator nav, TypeDefinition type)
```

Comment:
Why so many tiny methods - For example we can easily merge `ProcessSelectedFields` and `ProcessFields` into one - no reason to keep them as separate methods.